java/vector/src/main/java/org/apache/arrow/vector/DecimalVector.java

```diff
@@ -206,41 +207,30 @@ public void set(int index, ArrowBuf buffer) {
   public void setBigEndian(int index, byte[] value) {
     BitVectorHelper.setValidityBitToOne(validityBuffer, index);
     final int length = value.length;
-    int startIndex = index * TYPE_WIDTH;
-    if (length == TYPE_WIDTH) {
-      for (int i = TYPE_WIDTH - 1; i >= 3; i -= 4) {
-        valueBuffer.setByte(startIndex, value[i]);
-        valueBuffer.setByte(startIndex + 1, value[i - 1]);
-        valueBuffer.setByte(startIndex + 2, value[i - 2]);
-        valueBuffer.setByte(startIndex + 3, value[i - 3]);
-        startIndex += 4;
-      }
 
-      return;
+    // do the bound check.
+    valueBuffer.checkBytes(index * TYPE_WIDTH, (index + 1) * TYPE_WIDTH);
+
+    long outAddress = valueBuffer.memoryAddress() + index * TYPE_WIDTH;
+    // swap bytes to convert BE to LE
+    for (int byteIdx = 0; byteIdx < length; ++byteIdx) {
+      PlatformDependent.putByte(outAddress + byteIdx, value[length - 1 - byteIdx]);
```

Comment:
I think the performance improvements are mainly because there is no boundary checking in PlatformDependent.putByte. So the same effects can be obtained by disabling the flag in org.apache.arrow.memory#BoundsChecking

Reply:
completely disabling the bounds check is a risk. my change preserves the bounds check but does it once per the setXX api rather than once for every input and output byte (i.e a 32x reduction).

Reply:
I see. I agree that it is a good idea to do boundary check once at the beginning.